OSSM-3180: Integration of cert-manager with Service-Mesh (#57747)
openshift/openshift-docs: modules/ossm-install-cert-manager.adoc

// Module included in the following assemblies:
//
// * service_mesh/v2x/ossm-security.adoc

:_content-type: PROCEDURE
[id="ossm-install-cert-manager_{context}"]
== Install cert-manager

.Procedure

. Create root cluster issuer:
+
[source, terminal]
----
oc apply -f cluster-issuer.yaml
----
+
[source, yaml]
----
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: selfsigned-root-issuer

>   namespace: openshift-operators

Reviewer: ```suggestion
  namespace: cert-manager
```

spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: root-ca

>   namespace: openshift-operators

Reviewer: ```suggestion
  namespace: cert-manager
```

spec:
  isCA: true
  duration: 21600h # 900d
  secretName: root-ca
  commonName: root-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: selfsigned-root-issuer
    kind: Issuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: root-ca
spec:
  ca:
    secretName: root-ca
----
+
====
[NOTE]

> The namespace of `selfsigned-root-issuer` issuer and `root-ca` certificate is `openshift-operators`. This is necessary because `root-ca` is a cluster issuer, so cert-manager will look for a referenced secret in its own namespace, which is `openshift-operators` in this case.

Reviewer: ```suggestion
The namespace of `selfsigned-root-issuer` issuer and `root-ca` certificate is `cert-manager`. This is necessary because `root-ca` is a cluster issuer, so cert-manager will look for a referenced secret in its own namespace, which is `cert-manager` in case of Red Hat cert-manager.
```

Author: Is "of Red Hat cert-manager" referring to "cert-manager Operator for Red Hat OpenShift"?

====

> . Create an intermediate CA certificate for istiod:

Reviewer: This section is irrelevant for istio-csr. Please take a look at these instructions once again: https://github.com/maistra/istio-operator/tree/maistra-2.4/deploy/examples/cert-manager/istio-csr

Author: I think I got myself confused, so I want to check. 

Do I replace theses steps:

Create an intermediate CA certificate for istiod:
Deploy SMCP:
Deploy apps:

with these:

Install cert-manager istio-csr:
Deploy Istio:
Deploy bookinfo app:

and then do i replace 

Test mTLS between apps: 

with

Test traffic:

i think i was looking at `custom-cert-rotation/README.md`file as the title is also "Integration with cert-manager and istio-csr", and didn't realize there was `istio-csr/README.md` so I want to check that I'm updating and adding the correct information. 

Which brings to another questions: do I need to create docs for `custom-cert-rotation` steps? Or are those no longer relevant?

Reviewer 2: Steps:

1. Install the `cert-manager Operator for Red Hat OpenShift ` from the OCP cluster
2. Create the rootCA certificate on the namespace cert-manager

3. Create the istioCA certificate on the namespace Istio-system
4. Install Istio-CSR on the namespace Istio-system 
5. Deploy SMCP

Reviewer 2: Steps:

Install the cert-manager Operator for Red Hat OpenShift  from the OCP cluster

Create the rootCA certificate on the namespace cert-manager

```
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: selfsigned-root-issuer
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: selfsigned-ca
spec:
  isCA: true
  duration: 21600h # 900d
  secretName: root-ca
  commonName: root-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: selfsigned-root-issuer
    kind: Issuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: root-ca
spec:
  ca:
    secretName: root-ca
```

Create the istioCA certificate on the namespace Istio-system

Install Istio-CSR on the namespace Istio-system

Deploy SMCP

Reviewer 2: Steps:

1. Install the cert-manager Operator for Red Hat OpenShift on the  NS cert-manager-operator
2. Apply the rootCA on the NS `cert-manager`

```
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: selfsigned-root-issuer
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: selfsigned-ca
spec:
  isCA: true
  duration: 21600h # 900d
  secretName: root-ca
  commonName: root-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: selfsigned-root-issuer
    kind: Issuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: root-ca
spec:
  ca:
    secretName: root-ca

```
3. Apply the istioCA on the NS istio-system

```
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: istio-ca
spec:
  isCA: true
  duration: 21600h
  secretName: istio-ca
  commonName: istio-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: root-ca
    kind: ClusterIssuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: istio-ca
spec:
  ca:
    secretName: istio-ca
```

4. Install the istio-csr on `istio-system`

5. deploy the SMCP on `istio-system
`

```
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
spec:
  security:
    certificateAuthority:
      cert-manager:
        address: cert-manager-istio-csr.%s.svc:443
      type: cert-manager
    dataPlane:
      mtls: true
    identity:
      type: ThirdParty
  tracing:
    type: None
  version: 2.4
```

then Verification

Reviewer 2: cc @dymurray @jewertow :

Review these steps

+
[source, terminal]
----
oc new-project istio-system

> oc apply -f cacerts.yaml
> ----
> +
> [source, yaml]
> ----
> apiVersion: cert-manager.io/v1
> kind: Certificate
> metadata:
>   name: cacerts
>   namespace: istio-system
> spec:
>   secretName: cacerts
>   duration: 1440h
>   renewBefore: 360h
>   commonName: istiod-basic.istio-system.svc
>   isCA: true
>   usages:
>   - digital signature
>   - key encipherment
>   - cert sign
>   dnsNames:
>   - istiod-basic.istio-system.svc
>   issuerRef:
>     name: root-ca
>     kind: ClusterIssuer
>     group: cert-manager.io

Reviewer: ```suggestion
oc apply -n istio-system -f istio-ca.yaml
----
+
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: istio-ca
  namespace: istio-system
spec:
  isCA: true
  duration: 21600h
  secretName: istio-ca
  commonName: istio-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: root-ca
    kind: ClusterIssuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: istio-ca
  namespace: istio-system
spec:
  ca:
    secretName: istio-ca
```

----

. Deploy SMCP:
+
[source, terminal]
----
oc apply -f mesh.yaml -n istio-system
----
+
[source, yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
spec:
  addons:
    grafana:
      enabled: false
    kiali:
      enabled: false
    prometheus:
      enabled: false

>   gateways:
>     ingress:
>       routeConfig:
>         enabled: false
>     egress:
>       enabled: false
>     openshiftRoute:
>       enabled: false

Reviewer: Please remove custom gateway configs.
```suggestion
```

  proxy:
    accessLogging:
      file:
        name: /dev/stdout

>   runtime:
>     components:
>       pilot:
>         container:
>           env:
>             AUTO_RELOAD_PLUGIN_CERTS: "true"

Reviewer: This is not relevant to istio-csr. Please remove it.
```suggestion
```

>   security:
>     dataPlane:
>       mtls: true

Reviewer: You forgot about the most important part of configuration:
```suggestion
security:
    certificateAuthority:
      cert-manager:
        address: cert-manager-istio-csr.istio-system.svc:443
      type: cert-manager
    dataPlane:
      mtls: true
    identity:
      type: ThirdParty
```

// Module included in the following assemblies:
//
// * service_mesh/v2x/ossm-security.adoc

:_content-type: PROCEDURE
[id="ossm-install-cert-manager_{context}"]
== Install cert-manager

> To install cert-manager, follow these steps:

Reviewer: Please update the steps below based on my new PR (look at istio-csr directory): https://github.com/maistra/istio-operator/pull/1156. These samples were tested with cert-manager operator and istio-csr v0.6.0.

oc apply -n istio-system -f istio-ca.yaml
----
+
[source, yaml]
----
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: selfsigned-root-issuer
  namespace: cert-manager
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: root-ca
  namespace: cert-manager
spec:
  isCA: true
  duration: 21600h # 900d
  secretName: root-ca
  commonName: root-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: selfsigned-root-issuer
    kind: Issuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: root-ca
spec:
  ca:
    secretName: root-ca
----
+
[source, yaml]
----
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: istio-ca
  namespace: istio-system
spec:
  isCA: true
  duration: 21600h
  secretName: istio-ca
  commonName: istio-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: root-ca
    kind: ClusterIssuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: istio-ca
  namespace: istio-system
spec:
  ca:
    secretName: istio-ca
----
+
====
[NOTE]

> The namespace of `selfsigned-root-issuer` issuer and `root-ca` certificate is `cert-manager`. This is necessary because `root-ca` is a cluster issuer, so cert-manager will look for a referenced secret in its own namespace, which is `cert-manager` in the case of Red Hat.

Reviewer: ```suggestion
The namespace of `selfsigned-root-issuer` issuer and `root-ca` certificate is `cert-manager`. This is necessary because `root-ca` is a cluster issuer, so cert-manager will look for a referenced secret in its own namespace, which is `cert-manager` in the case of Red Hat cert-manager.
```

  repository: quay.io/jetstack/cert-manager-istio-csr
  tag: v0.6.0
  pullSecretName: ""

app:
  certmanager:
    namespace: istio-system
    issuer:
      group: cert-manager.io
      kind: Issuer
      name: istio-ca

  controller:
    configmapNamespaceSelector: "maistra.io/member-of=istio-system"
    leaderElectionNamespace: istio-system

  istio:
    namespace: istio-system
    revisions: ["basic"]

  server:
    maxCertificateDuration: 5m

  tls:
    certificateDNSNames:
    # This DNS name must be set in the SMCP spec.security.certificateAuthority.cert-manager.address
    - cert-manager-istio-csr.istio-system.svc
----

. Deploy SMCP:
+
[source, terminal]
----
oc apply -f mesh.yaml -n istio-system
----
+
[source, yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
spec:
  addons:
    grafana:
      enabled: false
    kiali:
      enabled: false
    prometheus:
      enabled: false
  proxy:
    accessLogging:
      file:
        name: /dev/stdout
  security:
    certificateAuthority:
      cert-manager:
        address: cert-manager-istio-csr.istio-system.svc:443
      type: cert-manager
    dataPlane:
      mtls: true
    identity:

>       type: ThirdParty

Reviewer: I would also add a note below that `security.identity.type: ThirdParty` must be set when `security.certificateAuthority.type: cert-manager` is configured.

//
// * service_mesh/v2x/ossm-security.adoc

:_content-type: PROCEDURE
[id="ossm-install-cert-manager_{context}"]
== Install cert-manager

.Procedure

. Create root cluster issuer:
+
[source, terminal]
----
oc apply -f cluster-issuer.yaml
----
+
[source, yaml]
----
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: selfsigned-root-issuer
  namespace: cert-manager
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: root-ca
  namespace: cert-manager
spec:
  isCA: true
  duration: 21600h # 900d
  secretName: root-ca
  commonName: root-ca.my-company.net
  subject:
    organizations:
    - my-company.net
  issuerRef:
    name: selfsigned-root-issuer
    kind: Issuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: root-ca
spec:
  ca:
    secretName: root-ca
----
+
====
[NOTE]
The namespace of `selfsigned-root-issuer` issuer and `root-ca` certificate is `cert-manager`. This is necessary because `root-ca` is a cluster issuer, so cert-manager will look for a referenced secret in its own namespace, which is `cert-manager` in the case of Red Hat.
====

> . Install cert-manager `istio-csr`:

Reviewer: ```suggestion
Install `istio-csr`:
```

    organizations:
    - my-company.net
  issuerRef:
    name: root-ca
    kind: ClusterIssuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: istio-ca
  namespace: istio-system
spec:
  ca:
    secretName: istio-ca
----
+
[source, yaml]
----
replicaCount: 2

image:
  repository: quay.io/jetstack/cert-manager-istio-csr
  tag: v0.6.0
  pullSecretName: ""

app:
  certmanager:
    namespace: istio-system
    issuer:
      group: cert-manager.io
      kind: Issuer
      name: istio-ca

  controller:
    configmapNamespaceSelector: "maistra.io/member-of=istio-system"
    leaderElectionNamespace: istio-system

  istio:
    namespace: istio-system
    revisions: ["basic"]

  server:
    maxCertificateDuration: 5m

  tls:
    certificateDNSNames:
    # This DNS name must be set in the SMCP spec.security.certificateAuthority.cert-manager.address
    - cert-manager-istio-csr.istio-system.svc
----

. Deploy SMCP:
+
[source, terminal]
----
oc apply -f mesh.yaml -n istio-system
----
+
[source, yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
spec:
  addons:
    grafana:
      enabled: false
    kiali:

>       enabled: false

Reviewer: Please disable Prometheus as well.
```suggestion
      enabled: false
    prometheus:
      enabled: false
```

Reviewer: You resolved it, but it's still not done. Please disable Prometheus.

>   proxy:

Reviewer: I know that I suggested to deploy gateways, but now I think that apps will be sufficient for this example.
```suggestion
  gateways:
    enabled: false
  proxy:
```

    accessLogging:
      file:
        name: /dev/stdout
  security:
    certificateAuthority:
      cert-manager:
        address: cert-manager-istio-csr.istio-system.svc:443
      type: cert-manager
    dataPlane:
      mtls: true
    identity:
      type: ThirdParty
  tracing:
    type: None
---
apiVersion: maistra.io/v1
kind: ServiceMeshMemberRoll
metadata:
  name: default
spec:
  members:

>   - bookinfo

Reviewer: ```suggestion
  - httpbin
  - sleep
```

----

. Deploy apps:
+
[source, terminal]
----
oc new-project sleep
oc apply -f https://raw.githubusercontent.com/maistra/istio/maistra-2.4/samples/sleep/sleep.yaml -n sleep
oc new-project httpbin
oc apply -f https://raw.githubusercontent.com/maistra/istio/maistra-2.4/samples/httpbin/httpbin.yaml -n httpbin
----

> . Test traffic:

Reviewer: ```suggestion
. Test mTLS traffic between applications:
```

+
[source, terminal]
----

> while true; do curl -v bookinfo.apps-crc.testing:80/productpage > /dev/null; sleep 1; done

Reviewer: ```suggestion
oc exec $(oc get pods -n sleep -l app=sleep -o jsonpath='{.items[0].metadata.name}') -n sleep -c sleep -- \
    curl -v http://httpbin.httpbin.svc:8000/headers
```

// Module included in the following assemblies:
//
// * service_mesh/v2x/ossm-security.adoc

:_content-type: PROCEDURE
[id="ossm-install-cert-manager_{context}"]
== Install cert-manager

To install cert-manager, follow these steps:

//. Install cert-manager:
//+
//[source, terminal]
//----
//helm install cert-manager jetstack/cert-manager \
//    --namespace cert-manager \
//    --create-namespace \
//    --version v1.11.0 \
//    --set installCRDs=true
//----

. Provision certificates:
+
[source, terminal]
----
oc new-project istio-system

> oc apply -f deploy/examples/cert-manager/selfsigned-ca.yaml -n istio-system

Reviewer 2: Adding the Yaml will be clear, path will not clear for the Customers

Command:

`oc apply -f  <selfsigned-ca.yaml> -n istio-system
`
```
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: selfsigned
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: istio-ca
spec:
  isCA: true
  duration: 2160h # 90d
  secretName: istio-ca
  commonName: istio-ca
  subject:
    organizations:
      - cluster.local
      - cert-manager
  issuerRef:
    name: selfsigned
    kind: Issuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: istio-ca
spec:
  ca:
    secretName: istio-ca
```

Reviewer 3: +1

----

. Install cert-manager `istio-csr` service:
+
[source, terminal]
----
helm install -n cert-manager cert-manager-istio-csr \

>     jetstack/cert-manager-istio-csr -f deploy/examples/cert-manager/istio-csr-helm-values.yaml

Reviewer 4: This line references another yaml file from our example on the istio-operator repo. @pbajjuri20, shouldn't we include the content of this yaml as well?

Reviewer 2: Yes, It is also required the specific yaml

----

. Deploy Istio:
+
[source, terminal]
----

> oc apply -f deploy/examples/cert-manager/smcp.yaml -n istio-system

Reviewer 2: replace the Path with Yaml file

Author: To clarify: does `deploy/examples/cert-manager/smpc.yaml` get replaced with `<selfsigned-ca.yaml>`?
